### PR TITLE
Feat/add search by title

### PR DESCRIPTION
**Pull request by ephremAddiss**
We, at [Addis](https://addissoftware.com/), are integrating 1Password to our enterprise application and we want to add a feature that search all items that contains a given title value.

### CODE REVIEW

**Comment by volodymyrZotov** [comment]
@ephremAddiss Thank you very much for the contribution and for helping us to become better. 👍 
Such an ability is already done within [here](https://github.com/1Password/connect-sdk-js/pull/70).

It will be available in the next release. We are going to make it soon.

**Comment by ephremAddiss** [comment]
@volodymyrZotov I have checked the new release and i am afraid it doesn't include the functionality i proposed. It uses title exact match, what i am proposing is contains based title. It would make more sense to search for certain values in the title than remembering exact matching.

Could you please have a look at it?

Thanks

**Comment by volodymyrZotov** [comment]
@ephremAddiss Hello. 
Ok, I see. I'll take a closer look into the PR, thanks.

File: __test__/op-connect.test.ts [hunk 1]
```diff
@@ -458,6 +458,49 @@ describe("Test OnePasswordConnect CRUD", () => {
             expect(otp).toEqual(OTP);
         });
     });
+
+    describe("List items by title search", () => {
+        const title = "some title";
+        // Search all items by title which contains  the given value `title`
+        const getItemsByTitleMock = (title: string) => nock(mockServerUrl)
```
Comment (volodymyrZotov) [inline __test__/op-connect.test.ts:465]: You can use the `apiMock.listItemsByTitleSearch` method that you've created instead.

File: __test__/op-connect.test.ts [hunk 2]
```diff
@@ -458,6 +458,49 @@ describe("Test OnePasswordConnect CRUD", () => {
...
+                .get(`/v1/vaults/${VAULT_ID}/items/`)
+                .query({
+                    filter: `title co "${title}"`,
+                });
+        const getFullItemMock = (itemId: string) =>
```
Comment (volodymyrZotov) [inline __test__/op-connect.test.ts:470]: you can use `apiMock.getItemById` instead

File: src/lib/resources.ts [hunk 3]
```diff
@@ -258,6 +258,28 @@ export class Items extends OPResource {
...
 
+     /**
+     * Search by title for all Items which contains a given title.
+     *  If found, queries for complete item details and returns results
+     *
+     * @param {string} vaultId
+     * @param {string} itemTitle
+     * @returns {Promise<FullItem[]>}
+     */
+      public async listItemsByTitleSearch(
```
Comment (volodymyrZotov) [inline src/lib/resources.ts:269]: Let's rename to `listItemsByTitleContains`

File: src/lib/resources.ts [hunk 4]
```diff
@@ -258,6 +258,28 @@ export class Items extends OPResource {
         return ObjectSerializer.deserialize(data[0], "Item");
     }
 
+     /**
+     * Search by title for all Items which contains a given title.
+     *  If found, queries for complete item details and returns results
+     *
+     * @param {string} vaultId
+     * @param {string} itemTitle
```
Comment (volodymyrZotov) [inline src/lib/resources.ts:266]:
```suggestion
     * @param {string} titleSearchStr
```

File: src/lib/op-connect.ts [hunk 5]
```diff
@@ -175,6 +175,23 @@ class OPConnect {
...
+     *
+     * The Item Title doesn't have to match exactly instead must contain some value.
+     *
+     * @param {string} vaultId
+     * @param {string} title
+     * @returns {Promise<FullItem[]>}
+     */
+     public async listItemsByTitleSearch(
+        vaultId: string,
+        title: string,
```
Comment (volodymyrZotov) [inline src/lib/op-connect.ts:190]:
```suggestion
        titleSearchStr: string,
```

File: src/lib/op-connect.ts [hunk 6]
```diff
@@ -175,6 +175,23 @@ class OPConnect {
...
     }
 
+
+    /**
+     * Get details about all Item which contains a given Title value.
+     *
+     * The Item Title doesn't have to match exactly instead must contain some value.
+     *
+     * @param {string} vaultId
+     * @param {string} title
```
Comment (volodymyrZotov) [inline src/lib/op-connect.ts:185]:
```suggestion
     * @param {string} titleSearchStr
```

File: src/lib/resources.ts (hunk 3 above)
Comment (volodymyrZotov) [reply src/lib/resources.ts:269]: Please put this method below `listItemsByTitle`

File: src/lib/resources.ts [hunk 7]
```diff
@@ -258,6 +258,28 @@ export class Items extends OPResource {
         return ObjectSerializer.deserialize(data[0], "Item");
     }
 
+     /**
+     * Search by title for all Items which contains a given title.
```
Comment (volodymyrZotov) [inline src/lib/resources.ts:262]:
```suggestion
     * Search for the Items in which the Title contains a provided string.
```

File: src/lib/resources.ts [hunk 8]
```diff
@@ -258,6 +258,28 @@ export class Items extends OPResource {
...
+     *  If found, queries for complete item details and returns results
```
Comment (volodymyrZotov) [inline src/lib/resources.ts:263]:
```suggestion
```

File: src/lib/utils/query-builder.ts [hunk 9]
```diff
@@ -8,9 +8,13 @@ export const FILTER_PARAM = {
...
+const coByTitle = (title: string): string => `${FILTER_PARAM.TITLE} co "${title}"`;
+
 const buildKeyValuePair = (queryParamName: string): (queryParamValue: string) => string =>
     (queryParamValue): string => `${queryParamName}=${queryParamValue}`;
 
 const buildFilterQuery = buildKeyValuePair(QUERY_PARAM_NAME.FILTER);
 
 export const filterByTitle = (title: string): string => buildFilterQuery(eqByTitle(title));
+
+export const searchByTitle = (title: string): string => buildFilterQuery(coByTitle(title));
```
Comment (volodymyrZotov) [inline src/lib/utils/query-builder.ts:20]: Please add an empty line at the end of the file.

**Review by volodymyrZotov — CHANGES_REQUESTED** [review]

**Comment by ephremAddiss** [comment]
Done. And @volodymyrZotov Thanks for all the suggestions and improvement.

**Review by ephremAddiss — COMMENTED** [review]
@volodymyrZotov  I have resolved the changes you have requested. Have a look and give me a feedback.
Thanks.

File: __test__/op-connect.test.ts [hunk 10]
```diff
@@ -458,6 +458,43 @@ describe("Test OnePasswordConnect CRUD", () => {
             expect(otp).toEqual(OTP);
         });
     });
+
+    describe("List items by title contains", () => {
+        const title = "some title";
+        // Search all items by title which contains  the given value `title`
```
Comment (volodymyrZotov) [inline __test__/op-connect.test.ts:464]:
```suggestion
```

Comment (volodymyrZotov) [reply __test__/op-connect.test.ts:464]: @ephremAddiss Please remove this comment. I think the test case description is self-descriptive enough.

**Comment by volodymyrZotov** [comment]
Hi @ephremAddiss . Well done! Looks very good. Thank you! 👍

**Review by volodymyrZotov — APPROVED** [review]

**Comment by volodymyrZotov** [comment]
@ephremAddiss Hi. I've just noticed that your commits are not signed. Please sign them in order to be able to merge.
You can create another PR with the same changes and signed commits. Or do it in a way that's easier for you.

**Comment by ephremAddiss** [comment]
@volodymyrZotov  Hello, I have used `git rebase` to sign my prev commits and i have managed to change some but other commits are hard to change since i have merged some changes from your previous release.

Is it possible to merge on the current state or should i start from scratch? thanks

**Comment by volodymyrZotov** [comment]
@ephremAddiss I can't merge if there are unsigned commits.

**Comment by ephremAddiss** [comment]
Okay, I will start from scratch.

**Comment by volodymyrZotov** [comment]
Implemented here #82 as this branch contains unsigned commits.

**Pull request closed without merging**